### DISCUSSION TITLE
Advice on when/how to define and set workBounds

### DISCUSSION THREAD

**Discussion by EdenRavecca** [bmcclintock/momentuHMM#180] [Q&A] [answered]
Hello, I’m aiming to analyze movement data (GPS fixes at 5-second intervals) for 17 individuals, using an HMM with four data streams (step length, turning angle, altitude, change in altitude) and four states.

I would like to set constraints for the data streams, but I’m having difficulty adapting example 3.7 of the momentuHMM vignette to my data. 

As an example, I created a design matrix (DM) constraining mean step length, such that:
state1 > state2 & state3 and state4 > state2 & state3

```
slDM <- matrix(c (1,1,1,0,0,0,0,0, # state1 SL mean: 1>2, 1>3
                  0,1,0,0,0,0,0,0,
                  0,0,1,0,0,0,0,0,
                  0,1,1,1,0,0,0,0, # state4 SL mean: 4>2, 4>3
                  0,0,0,0,1,0,0,0,
                  0,0,0,0,1,1,1,1, # state2 SL sd: 2>1, 2>3, 2>4
                  0,0,0,0,0,0,1,0,
                  0,0,0,0,0,0,0, 1),
               # nrow = number of states * number of params in distribution
               nrow = nbStates*2, byrow = TRUE,
               dimnames=list(c(paste0("mean_",1:nbStates),
                               paste0("sd_",1:nbStates)),
                             c(paste0("mean_",1:nbStates,"(Int)"),
                               paste0("sd_",1:nbStates,"(Int)"))) )
```

For userBounds, I set lower and upper step length values on the natural scale:

```
slBounds <- matrix( c(40, 320, # state1 mean bounds
                      0.33, 100, # state2
                      0.33, 80, # state3
                      0.33, 320,# state4
                      0,10, # state1 sd bounds
                      0,20, # state2 
                      0,15, # state3
                      0,10 ), # state4
                    nrow = nbStates*2, byrow = TRUE,
                    dimnames = list( rownames(slDM), c("lower", "upper") ) )
```

I am stuck on the process of setting workBounds or transforming userBound values from the natural scale to a working scale. Specifically, understanding what the workBounds are meant to be for different distributions and when workBounds are required or need to be modified. 

For the step length example, given a gamma distribution with a mean and standard deviation, I am assuming that the mean ranges between -Inf and Inf, and the standard deviation is between 0 and Inf. Is this correct? If I set lower and upper bounds with slBounds, do I need to set workBounds/alter the ones below? If so, how? 

Currently they are:
```
slworkBounds <- matrix(c(rep(-Inf, nbStates), rep(0, nbStates),
                         rep(Inf, nbStates*2) ), nrow = nbStates*2,byrow = FALSE,
                         dimnames = list(colnames(slDM), c("lower", "upper"))  )
```

I would greatly appreciate any guidance you could offer or direction to who I may be able to chat with on how to define workBounds for different distributions and how to alter them if I only modify the DM, or if I only modify the realBounds, or if I modify both, as in the step example I provided.

Thanks!

**Comment by bmcclintock** [comment, marked as answer]
Using both `userBounds` and `workBounds` relational constraints can be difficult and have unintended consequences (owing to the different scales of the parameters), but it is demonstrated in the [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf) examples (including the equations used). I'd suggest starting with just `workBounds` to enforce relational constraints and see if `userBounds` constraints are really necessary. If you still want to use both, I'd suggest carefully going through the equations and verifying that you're imposing the desired constraints on the real scale.

The real parameter bounds and link functions are listed in Table 2 of the [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf). It also looks like you have the rows and columns transposed in `slDM`. See #42. 

One cannot constrain 1>2, 1>3, 4>2, and 4>3 without imposing some relationship between 1 and 4 (or 2 and 3). For example:

```
slDM <- matrix(c (1,0,0,0,0,0,0,0, # state1 SL mean: 1>2, 1>3
                  1,1,0,1,0,0,0,0,
                  1,0,1,1,0,0,0,0,
                  0,0,0,1,0,0,0,0, # state4 SL mean: 4>2, 4>3
                  0,0,0,0,1,1,0,0,
                  0,0,0,0,0,1,0,0, # state2 SL sd: 2>1, 2>3, 2>4
                  0,0,0,0,0,1,1,0,
                  0,0,0,0,0,1,0,1),
               # nrow = number of states * number of params in distribution
               nrow = nbStates*2, byrow = TRUE,
               dimnames=list(c(paste0("mean_",1:nbStates),
                               paste0("sd_",1:nbStates)),
                             c(paste0("mean_",1:nbStates,"(Int)"),
                               paste0("sd_",1:nbStates,"(Int)"))) )

slworkBounds <- matrix(c(rep(-Inf,4),rep(-Inf,4),Inf,0,0,Inf,0,Inf,0,0), nrow = nbStates*2,byrow = FALSE,
                                               dimnames = list(colnames(slDM), c("lower", "upper"))  )
```
gives $$mean_1=\exp(\beta_1)$$ $$mean_2 = \exp(\beta_1 - (\exp(-\beta_2)-0) + \beta_4)$$ $$mean_3 = \exp(\beta_1 - (\exp(-\beta_3)-0) + \beta_4)$$ $$mean_4=\exp(\beta_4)$$ $$sd_1 = \exp(-(\exp(-\beta_5)-0)+\beta_6)$$ $$sd_2=\exp(\beta_6)$$ $$sd_3=\exp(\beta_6-(\exp(-\beta_7)-0))$$ $$sd_4=\exp(\beta_6-(\exp(-\beta_8)-0))$$ which does not guarantee 1>2, 1>3, 4>2, or 4>3 for the mean parameter.
